### PR TITLE
Rework sparse streams, verify DMG images

### PR DESCRIPTION
**Pull request by zvin**
Change-type: major

### CODE REVIEW

File: lib/multi-write.ts [hunk 1]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2018 balena.io
+ * Copyright 2019 balena.io
```
Comment (thundron) [inline lib/multi-write.ts]: Why though? If the file's been *created* in 2018, it's fine

Comment (zvin) [reply lib/multi-write.ts]: I updated the files I modified, switched back to the previous value.

File: lib/source-destination/configured-source/configured-source.ts [hunk 2]
```diff
@@ -156,28 +159,24 @@ export class ConfiguredSource extends SourceSource {
...
-			this.read.bind(this),
-			await this.getBlockmap(),
-			false,
-			generateChecksums,
-			0,
-			Infinity,
+	): Promise<SparseReadStream> {
+		return new SparseReadStream(
+			this,
+			await this.getBlocksWithChecksumType(generateChecksums),
```
[the remark this answers is not in the document]
Comment (zvin) [reply lib/source-destination/configured-source/configured-source.ts]: comment added

File: lib/errors.ts [hunk 3]
```diff
@@ -31,6 +33,28 @@ export class ChecksumVerificationError extends VerificationError {
 		super(message);
 	}
 }
+
+function blocksVerificationErrorMessage(
+	blocks: BlocksWithChecksum,
+	checksum: string,
+) {
+	const start = blocks.blocks[0].offset;
```
[the remark this answers is not in the document]
Comment (zvin) [reply lib/errors.ts]: updated

**Pull request merged**